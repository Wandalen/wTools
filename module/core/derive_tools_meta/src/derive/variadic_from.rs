
use super::*;
use macro_tools::{ Result, format_ident, attr, diag };
use iter::{ IterExt, Itertools };

/// This function generates an implementation of a variadic `From` trait for a given struct.
/// It handles both named and unnamed fields within the struct, generating appropriate code
/// for converting a tuple of fields into an instance of the struct.

pub fn variadic_from( input : proc_macro::TokenStream ) -> Result< proc_macro2::TokenStream >
{

  let original_input = input.clone();
  let parsed = syn::parse::< syn::ItemStruct >( input )?;
  let has_debug = attr::has_debug( parsed.attrs.iter() )?;
  let item_name = &parsed.ident;

  let len = parsed.fields.len();
  let from_trait = format_ident!( "From{len}",  );
  let from_method = format_ident!( "from{len}" );
  // xxx : test for zero fields

  let
  (
    types,
    fn_params,
    src_into_vars,
    vars
  )
  :
  ( Vec< _ >, Vec< _ >, Vec< _ >, Vec< _ > )
  = parsed.fields.iter().enumerate().map_result( | ( i, field ) |
  {
    let ident = field.ident.clone().map_or_else( || format_ident!( "_{i}" ), | e | e );
    let ty = field.ty.clone();
    Result::Ok
    ((
      qt!{ #ty, },
      qt!{ #ident : #ty, },
      qt!{ let #ident = ::core::convert::Into::into( #ident ); },
      qt!{ #ident, },
    ))
  })?
  .into_iter()
  .multiunzip();

  let result = match &parsed.fields
  {
    syn::Fields::Named( _ ) =>
    {

      if 1 <= len && len <= 3
      {
        qt!
        {

          #[ automatically_derived ]
          // impl variadic_from::From2< i32 > for StructNamedFields
          impl variadic_from::#from_trait< #( #types )* > for #item_name
          {
            // fn from1( a : i32, b : i32 ) -> Self
            fn #from_method
            (
              #( #fn_params )*
            ) -> Self
            {
              #( #src_into_vars )*
              // let a = ::core::convert::Into::into( a );
              // let b = ::core::convert::Into::into( b );
              Self
              {
                #( #vars )*
                // a,
                // b,
              }
            }
          }

          impl From< ( #( #types )* ) > for #item_name
          {
            /// Reuse From1.
            #[ inline( always ) ]
            fn from( src : ( #( #types )* ) ) -> Self
            {
              Self::from1( src )
            }
          }

        }
      }
      else
      {
        qt!{}
      }

    }
    syn::Fields::Unnamed( _ ) =>
    {

      if 1 <= len && len <= 3
      {
        qt!
        {

          #[ automatically_derived ]
          // impl variadic_from::From2< i32 > for StructNamedFields
          impl variadic_from::#from_trait< #( #types )* > for #item_name
          {
            // fn from1( a : i32, b : i32 ) -> Self
            fn #from_method
            (
              #( #fn_params )*
            ) -> Self
            {
              #( #src_into_vars )*
              // let a = ::core::convert::Into::into( a );
              // let b = ::core::convert::Into::into( b );
              Self
              (
                #( #vars )*
                // a,
                // b,
              )
            }
          }

          impl From< ( #( #types )* ) > for #item_name
          {
            /// Reuse From1.
            #[ inline( always ) ]
            fn from( src : ( #( #types )* ) ) -> Self
            {
              Self::from1( src )
            }
          }

        }
      }
      else
      {
        qt!{}
      }

    }
    syn::Fields::Unit =>
    {

      qt!{}

    }
    // _ => return Err( syn_err!( parsed.fields.span(), "Expects fields" ) ),
  };

  if has_debug
  {
    let about = format!( "derive : VariadicForm\nstructure : {item_name}" );
    diag::report_print( about, &original_input, &result );
  }

  Ok( result )
}
