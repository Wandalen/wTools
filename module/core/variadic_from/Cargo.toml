[package]
name = "variadic_from"
version = "0.22.0"
edition = "2021"
authors = [
  "Kostiantyn Wandalen <wandalen@obox.systems>",
  "Dmytro Kryvoruchko <dmytro.kr@obox.systems>",
]
license = "MIT"
readme = "Readme.md"
documentation = "https://docs.rs/variadic_from"
repository = "https://github.com/Wandalen/wTools/tree/master/module/core/variadic_from"
homepage = "https://github.com/Wandalen/wTools/tree/master/module/core/variadic_from"
description = """
Variadic from.
"""
categories = [ "algorithms", "development-tools" ]
keywords = [ "fundamental", "general-purpose" ]

[lints]
workspace = true

[package.metadata.docs.rs]
features = [ "full" ]
all-features = false



[features]

default = [
  "enabled",
  "derive_variadic_from",
  "type_variadic_from",
]

full = [
  "enabled",
  "derive_variadic_from",
  "type_variadic_from",
]
no_std = []
use_alloc = [ "no_std" ]
enabled = []

type_variadic_from = []
derive_variadic_from = [ "type_variadic_from", "derive_tools_meta/derive_variadic_from" ]

[dependencies]

## internal
derive_tools_meta = { workspace = true, features = [ "enabled", "derive_variadic_from" ] }

[dev-dependencies]
test_tools = { workspace = true }
