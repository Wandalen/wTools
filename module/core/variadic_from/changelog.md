# Changelog

*   **2025-06-29:**
    *   Implemented the `VariadicFrom` derive macro and `from!` helper macro, adhering to `spec.md`. Defined `FromN` traits, added blanket `From1` implementations, implemented `from!` macro with argument count validation, and ensured the derive macro generates `FromN` and `From<T>`/`From<tuple>` implementations based on field count (1-3 fields). Removed `#[from(Type)]` attribute handling. All generated code compiles without errors, passes tests (including doc tests, with `Readme.md` examples now runnable), and adheres to `clippy` warnings. Improved `Readme.md` content and scaffolding for new developers.

*   **2025-07-01:**
    *   Generalized `CONTRIBUTING.md` to be about all crates of the `wTools` repository, including updating the title, removing specific crate paths, and generalizing commit message examples.

*   [2025-07-06] Refactored `variadic_from_meta` to align with spec v1.1.
