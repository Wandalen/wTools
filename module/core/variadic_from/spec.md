# Technical Specification: `variadic_from` Crate

### 1. Introduction & Core Concepts

#### 1.1. Goals & Philosophy

The primary goal of the `variadic_from` crate is to enhance developer ergonomics and reduce boilerplate code in Rust by providing flexible, "variadic" constructors for structs. The core philosophy is to offer a single, intuitive, and consistent interface for struct instantiation, regardless of the number of initial arguments (within defined limits).

The framework is guided by these principles:
*   **Convention over Configuration:** The system should work out-of-the-box with sensible defaults. The `VariadicFrom` derive macro should automatically generate the necessary implementations for the most common use cases without requiring manual configuration.
*   **Minimal Syntactic Noise:** The user-facing `from!` macro provides a clean, concise way to construct objects, abstracting away the underlying implementation details of which `FromN` trait is being called.
*   **Seamless Integration:** The crate should feel like a natural extension of the Rust language. It achieves this by automatically implementing the standard `From<T>` trait for single fields and `From<tuple>` for multiple fields, enabling idiomatic conversions like `.into()`.
*   **Non-Intrusive Extensibility:** While the derive macro handles the common cases, the system is built on a foundation of public traits (`From1`, `From2`, etc.) that developers can implement manually for custom behavior or to support types not covered by the macro.

#### 1.2. Key Terminology (Ubiquitous Language)

*   **Variadic Constructor:** A constructor that can accept a variable number of arguments. In the context of this crate, this is achieved through the `from!` macro.
*   **`FromN` Traits:** A set of custom traits (`From1`, `From2`, `From3`) that define a contract for constructing a type from a specific number (`N`) of arguments.
*   **`VariadicFrom` Trait:** A marker trait implemented via a derive macro (`#[derive(VariadicFrom)]`). Its presence on a struct signals that the derive macro should automatically implement the appropriate `FromN` and `From<T>`/`From<tuple>` traits based on the number of fields in the struct.
*   **`from!` Macro:** A declarative, user-facing macro that provides the primary interface for variadic construction. It resolves to a call to `Default::default()`, `From1::from1`, `From2::from2`, or `From3::from3` based on the number of arguments provided.
*   **Named Struct:** A struct where fields are defined with explicit names, e.g., `struct MyStruct { a: i32 }`.
*   **Unnamed Struct (Tuple Struct):** A struct where fields are defined by their type only, e.g., `struct MyStruct(i32)`.

#### 1.3. Versioning Strategy

The `variadic_from` crate adheres to the Semantic Versioning 2.0.0 (SemVer) standard.
*   **MAJOR** version changes indicate incompatible API changes.
*   **MINOR** version changes introduce new, backward-compatible functionality (e.g., increasing the maximum number of supported arguments).
*   **PATCH** version changes are for backward-compatible bug fixes.

This specification document is versioned in lockstep with the crate itself.

### 2. Core Object Definitions

This section provides the formal definitions for the traits that constitute the `variadic_from` framework. These traits define the contracts that are either implemented automatically by the derive macro or manually by the user.

#### 2.1. The `FromN` Traits

The `FromN` traits provide a standardized interface for constructing a type from a specific number (`N`) of arguments.

##### 2.1.1. `From1<Arg>`
*   **Purpose:** Defines a contract for constructing an object from a single argument. It also serves as a unified interface for converting from tuples of varying lengths, which are treated as a single argument.
*   **Signature:**
    ```rust
    pub trait From1<Arg>
    where
        Self: Sized,
    {
        fn from1(arg: Arg) -> Self;
    }
    ```
*   **Blanket Implementations:** The framework provides blanket implementations to unify tuple-based construction under `From1`:
    *   `impl<T, All> From1<(T,)> for All where All: From1<T>`
    *   `impl<T1, T2, All> From1<(T1, T2)> for All where All: From2<T1, T2>`
    *   `impl<T1, T2, T3, All> From1<(T1, T2, T3)> for All where All: From3<T1, T2, T3>`
    *   `impl<All> From1<()> for All where All: Default`

##### 2.1.2. `From2<Arg1, Arg2>`
*   **Purpose:** Defines a contract for constructing an object from exactly two arguments.
*   **Signature:**
    ```rust
    pub trait From2<Arg1, Arg2>
    where
        Self: Sized,
    {
        fn from2(arg1: Arg1, arg2: Arg2) -> Self;
    }
    ```

##### 2.1.3. `From3<Arg1, Arg2, Arg3>`
*   **Purpose:** Defines a contract for constructing an object from exactly three arguments.
*   **Signature:**
    ```rust
    pub trait From3<Arg1, Arg2, Arg3>
    where
        Self: Sized,
    {
        fn from3(arg1: Arg1, arg2: Arg2, arg3: Arg3) -> Self;
    }
    ```

#### 2.2. The `VariadicFrom` Trait

*   **Purpose:** This is a marker trait that enables the `#[derive(VariadicFrom)]` macro. It does not contain any methods. Its sole purpose is to be attached to a struct to signal that the derive macro should perform code generation for it.
*   **Definition:** The trait is defined externally (in `derive_tools_meta`) but is exposed through the `variadic_from` crate.
*   **Behavior:** When a struct is decorated with `#[derive(VariadicFrom)]`, the derive macro is responsible for:
    1.  Implementing the `VariadicFrom` trait for that struct.
    2.  Generating implementations for the appropriate `FromN` trait(s).
    3.  Generating an implementation for the standard `From<T>` trait (for single-field structs) or `From<tuple>` trait (for multi-field structs).

### 3. Processing & Execution Model

This section details the internal logic of the crate's two primary components: the `VariadicFrom` derive macro and the `from!` macro.

#### 3.1. The `VariadicFrom` Derive Macro

The derive macro is the core of the crate's code generation capabilities.

*   **Activation:** The macro is activated when a struct is annotated with `#[derive(VariadicFrom)]`.
*   **Processing Steps:**
    1.  The macro receives the Abstract Syntax Tree (AST) of the struct it is attached to.
    2.  It inspects the struct's body to determine its kind (Named or Unnamed/Tuple) and counts the number of fields.
    3.  It extracts the types of each field in their declared order.
*   **Code Generation Logic:**
    *   **If field count is 1, 2, or 3:**
        *   It generates an implementation of the corresponding `FromN` trait. For a struct with `N` fields, it generates `impl FromN<T1, ..., TN> for MyStruct`, where `T1..TN` are the field types. The body of the generated function constructs an instance of the struct, mapping the arguments to the fields in order.
        *   For structs with 2 or 3 fields, it generates an implementation of the standard `From<(T1, ..., TN)>` trait. The body of this implementation delegates directly to the newly implemented `FromN` trait, calling `Self::fromN(...)`.
        *   For structs with 1 field, it generates an implementation of the standard `From<T>` trait (where `T` is the type of the single field). The body of this implementation delegates directly to the newly implemented `From1` trait, calling `Self::from1(...)`.
    *   **If field count is 0 or greater than 3:** The derive macro generates no code. This is a deliberate design choice to prevent unexpected behavior for unsupported struct sizes.

#### 3.2. The `from!` Macro

The `from!` macro provides a convenient, unified syntax for variadic construction. It is a standard `macro_rules!` macro that dispatches to the correct implementation based on the number of arguments provided at the call site.

*   **Resolution Rules:**
    *   `from!()` expands to `::core::default::Default::default()`. This requires the target type to implement the `Default` trait.
    *   `from!(arg1)` expands to `$crate::From1::from1(arg1)`.
    *   `from!(arg1, arg2)` expands to `$crate::From2::from2(arg1, arg2)`.
    *   `from!(arg1, arg2, arg3)` expands to `$crate::From3::from3(arg1, arg2, arg3)`.
    *   `from!(arg1, ..., argN)` where `N > 3` results in a `compile_error!`, providing a clear message that the maximum number of arguments has been exceeded.

### 4. Interaction Modalities

Users can leverage the `variadic_from` crate in two primary ways, both designed to be idiomatic Rust.

#### 4.1. Direct Instantiation via `from!`

This is the most direct and expressive way to use the crate. It allows for the creation of struct instances with a variable number of arguments.

*   **Example:**
    ```rust
    // Assumes MyStruct has two fields: i32, i32
    // and also implements Default and From1<i32>

    // Zero arguments (requires `Default`)
    let s0: MyStruct = from!();

    // One argument (requires manual `From1<i32>`)
    let s1: MyStruct = from!(10);

    // Two arguments (uses generated `From2`)
    let s2: MyStruct = from!(10, 20);
    ```

#### 4.2. Tuple Conversion via `From` and `Into`

By generating `From<tuple>` implementations, the derive macro enables seamless integration with the standard library's conversion traits.

*   **Example:**
    ```rust
    // Assumes MyStruct has two fields: i32, i32

    // Using From::from
    let s1: MyStruct = MyStruct::from((10, 20));

    // Using .into()
    let s2: MyStruct = (10, 20).into();

    // Using from! with a tuple (leverages the From1 blanket impl)
    let s3: MyStruct = from!((10, 20));
    ```

### 5. Cross-Cutting Concerns

#### 5.1. Error Handling Strategy

All error handling occurs at **compile time**, which is ideal for a developer utility crate.
*   **Invalid Argument Count:** Calling the `from!` macro with more than 3 arguments results in a clear, explicit `compile_error!`.
*   **Unsupported Struct Size:** The `VariadicFrom` derive macro will simply not generate code for structs with 0 or more than 3 fields. This will result in a subsequent compile error if code attempts to use a non-existent `FromN` implementation (e.g., "no method named `from2` found").
*   **Type Mismatches:** Standard Rust type-checking rules apply. If the arguments passed to `from!` do not match the types expected by the corresponding `FromN` implementation, a compile error will occur.

#### 5.2. Extensibility Model

The framework is designed to be extensible through manual trait implementation.
*   **Custom Logic:** Users can (and are encouraged to) implement `From1` manually to provide custom construction logic from a single value, as shown in the `variadic_from_trivial.rs` example.
*   **Overriding Behavior:** A manual implementation of a `FromN` trait will always take precedence over a generated one if both were somehow present.
*   **Supporting Larger Structs:** For structs with more than 3 fields, users can manually implement the `From<tuple>` trait to provide similar ergonomics, though they will not be able to use the `from!` macro for more than 3 arguments.

### 6. Known Limitations

*   **Argument Count Limit:** The `VariadicFrom` derive macro and the `from!` macro are hard-coded to support a maximum of **three** arguments/fields. There is no support for variadic generics beyond this limit.
*   **Type Inference:** In highly complex generic contexts, the compiler may require explicit type annotations (turbofish syntax) to resolve the correct `FromN` implementation. This is a general characteristic of Rust's type system rather than a specific flaw of the crate.

### 7. Appendices

#### A.1. Code Examples

##### Named Struct Example
```rust
use variadic_from::exposed::*;

#[derive(Debug, PartialEq, Default, VariadicFrom)]
struct UserProfile {
    id: u32,
    username: String,
}

// Manual implementation for a single argument
impl From1<&str> for UserProfile {
    fn from1(name: &str) -> Self {
        Self { id: 0, username: name.to_string() }
    }
}

// Usage:
let u1: UserProfile = from!(); // -> UserProfile { id: 0, username: "" }
let u2: UserProfile = from!("guest"); // -> UserProfile { id: 0, username: "guest" }
let u3: UserProfile = from!(101, "admin".to_string()); // -> UserProfile { id: 101, username: "admin" }
let u4: UserProfile = (102, "editor".to_string()).into(); // -> UserProfile { id: 102, username: "editor" }
```

##### Unnamed (Tuple) Struct Example
```rust
use variadic_from::exposed::*;

#[derive(Debug, PartialEq, Default, VariadicFrom)]
struct Point(i32, i32, i32);

// Usage:
let p1: Point = from!(); // -> Point(0, 0, 0)
let p2: Point = from!(1, 2, 3); // -> Point(1, 2, 3)
let p3: Point = (4, 5, 6).into(); // -> Point(4, 5, 6)
```

### 8. Meta-Requirements

This specification document must adhere to the following rules to ensure its clarity, consistency, and maintainability.
*   **Ubiquitous Language:** All terms defined in the `Key Terminology` section must be used consistently throughout this document and all related project artifacts.
*   **Naming Conventions:** All asset names (files, variables, etc.) must use `snake_case`.
*   **Mandatory Structure:** This document must follow the agreed-upon section structure. Additions must be justified and placed appropriately.

### 9. Deliverables

Working solution.

### 10. Conformance Check Procedure

The following checks must be performed to verify that an implementation of the `variadic_from` crate conforms to this specification.

1.  **Derive on 2-Field Named Struct:**
    *   **Action:** Apply `#[derive(VariadicFrom)]` to a named struct with 2 fields.
    *   **Expected:** The code compiles. `impl From2` and `impl From<(T1, T2)>` are generated.
2.  **Derive on 3-Field Unnamed Struct:**
    *   **Action:** Apply `#[derive(VariadicFrom)]` to an unnamed (tuple) struct with 3 fields.
    *   **Expected:** The code compiles. `impl From3` and `impl From<(T1, T2, T3)>` are generated.
3.  **`from!` Macro Correctness:**
    *   **Action:** Call `from!()`, `from!(a)`, `from!(a, b)`, and `from!(a, b, c)` on conforming types.
    *   **Expected:** All calls compile and produce the correct struct instances as defined by the `Default`, `From1`, `From2`, and `From3` traits respectively.
4.  **`from!` Macro Error Handling:**
    *   **Action:** Call `from!(a, b, c, d)`.
    *   **Expected:** The code fails to compile with an error message explicitly stating the argument limit has been exceeded.
5.  **Tuple Conversion Correctness (2-3 fields):**
    *   **Action:** Use `(a, b).into()` and `MyStruct::from((a, b))` on a derived 2-field struct.
    *   **Expected:** Both conversions compile and produce the correct struct instance.
6.  **Single-Field Conversion Correctness:**
    *   **Action:** Use `a.into()` and `MyStruct::from(a)` on a derived 1-field struct.
    *   **Expected:** Both conversions compile and produce the correct struct instance.
7.  **Derive on 4-Field Struct:**
    *   **Action:** Apply `#[derive(VariadicFrom)]` to a struct with 4 fields and attempt to call `from!(a, b, c, d)`.
    *   **Expected:** The code fails to compile with an error indicating that no `From4` trait or method exists, confirming the derive macro did not generate code.
8.  **Manual `From1` Implementation:**
    *   **Action:** Create a struct with `#[derive(VariadicFrom)]` and also provide a manual `impl From1<T> for MyStruct`.
    *   **Expected:** Calling `from!(t)` uses the manual implementation, demonstrating that user-defined logic can coexist with the derived logic.