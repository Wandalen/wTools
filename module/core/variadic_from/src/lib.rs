#![ cfg_attr( feature = "no_std", no_std ) ]
#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ doc( html_favicon_url = "https://raw.githubusercontent.com/Wandalen/wTools/alpha/asset/img/logo_v3_trans_square_icon_small_v2.ico" ) ]
#![ doc( html_root_url = "https://docs.rs/derive_tools/latest/derive_tools/" ) ]
#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/", "Readme.md" ) ) ]

#[ cfg( feature = "enabled" ) ]
pub mod variadic;

/// Namespace with dependencies.

#[ cfg( feature = "enabled" ) ]
pub mod dependency
{
  pub use ::derive_tools_meta;
}

#[ cfg( feature = "enabled" ) ]
#[ doc( inline ) ]
#[ allow( unused_imports ) ]
pub use own::*;

/// Own namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod own
{
  use super::*;
  #[ doc( inline ) ]
  pub use orphan::*;
  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use super::variadic::orphan::*;
}

/// Orphan namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod orphan
{
  use super::*;
  #[ doc( inline ) ]
  pub use exposed::*;

}

/// Exposed namespace of the module.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod exposed
{
  use super::*;
  #[ doc( inline ) ]
  pub use prelude::*;

  #[ doc( inline ) ]
  pub use ::derive_tools_meta::*;

}

/// Prelude to use essentials: `use my_module::prelude::*`.
#[ cfg( feature = "enabled" ) ]
#[ allow( unused_imports ) ]
pub mod prelude
{
  use super::*;

  #[ doc( inline ) ]
  #[ allow( unused_imports ) ]
  pub use super::variadic::prelude::*;
  // #[ doc( no_inline ) ]
  // pub use super::variadic;
  // #[ doc( no_inline ) ]
  // pub use ::derive_tools_meta::VariadicFrom;

}
