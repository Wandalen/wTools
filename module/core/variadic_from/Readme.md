# Module :: variadic_from

<!--{ generate.module_header.start() }-->
 [![experimental](https://raster.shields.io/static/v1?label=&message=experimental&color=orange)](https://github.com/emersion/stability-badges#experimental) [![rust-status](https://github.com/Wandalen/wTools/actions/workflows/module_variadic_from_push.yml/badge.svg)](https://github.com/Wandalen/wTools/actions/workflows/module_variadic_from_push.yml) [![docs.rs](https://img.shields.io/docsrs/variadic_from?color=e3e8f0&logo=docs.rs)](https://docs.rs/variadic_from) [![Open in Gitpod](https://raster.shields.io/static/v1?label=try&message=online&color=eee&logo=gitpod&logoColor=eee)](https://gitpod.io/#RUN_PATH=.,SAMPLE_FILE=module%2Fcore%2Fvariadic_from%2Fexamples%2Fvariadic_from_trivial.rs,RUN_POSTFIX=--example%20variadic_from_trivial/https://github.com/Wandalen/wTools) [![discord](https://img.shields.io/discord/872391416519737405?color=eee&logo=discord&logoColor=eee&label=ask)](https://discord.gg/m3YfbXpUUY)
<!--{ generate.module_header.end }-->

The variadic from is designed to provide a way to implement the From-like traits for structs with a variable number of fields, allowing them to be constructed from tuples of different lengths or from individual arguments. This functionality is particularly useful for creating flexible constructors that enable different methods of instantiation for a struct. By automating the implementation of traits crate reduces boilerplate code and enhances code readability and maintainability.

Currently it support up to 3 arguments. If your structure has more than 3 fields derive generates nothing. Also it supports tuple conversion, allowing structs to be instantiated from tuples by leveraging the `From` and `Into` traits for seamless conversion.

### Basic use-case.

<!-- qqq : for Petro : make this generator working -->
<!--{ example.use{ code : "variadic_from_trivial", hidden_code : "variadic_from_expanded", link : true, try_out : true } }-->
<!--{ example.use.end }-->

This example demonstrates the use of the `variadic_from` macro to implement flexible
constructors for a struct, allowing it to be instantiated from different numbers of
arguments or tuples. It also showcases how to derive common traits like `Debug`,
`PartialEq`, `Default`, and `VariadicFrom` for the struct.

```rust
#[ cfg( not( all(feature = "enabled", feature = "type_variadic_from", feature = "derive_variadic_from" ) ) ) ]
fn main(){}
#[ cfg( all( feature = "enabled", feature = "type_variadic_from", feature = "derive_variadic_from" ) )]
fn main()
{
  use variadic_from::exposed::*;

  // Define a struct `MyStruct` with fields `a` and `b`.
  // The struct derives common traits like `Debug`, `PartialEq`, `Default`, and `VariadicFrom`.
  #[ derive( Debug, PartialEq, Default, VariadicFrom ) ]
  // Use `#[ debug ]` to expand and debug generate code.
  // #[ debug ]
  struct MyStruct
  {
    a : i32,
    b : i32,
  }

  // Implement the `From1` trait for `MyStruct`, which allows constructing a `MyStruct` instance
  // from a single `i32` value by assigning it to both `a` and `b` fields.

  impl From1< i32 > for MyStruct
  {
    fn from1( a : i32 ) -> Self { Self { a, b : a } }
  }

  let got : MyStruct = from!();
  let exp = MyStruct { a : 0, b : 0 };
  assert_eq!( got, exp );

  let got : MyStruct = from!( 13 );
  let exp = MyStruct { a : 13, b : 13 };
  assert_eq!( got, exp );

  let got : MyStruct = from!( 13, 14 );
  let exp = MyStruct { a : 13, b : 14 };
  assert_eq!( got, exp );

  dbg!( exp );
  //> MyStruct {
  //>   a : 13,
  //>   b : 14,
  //> }

}
```

<details>
<summary>The code above will be expanded to this</summary>

```rust
#[ cfg( not( all(feature = "enabled", feature = "type_variadic_from" ) ) ) ]
fn main(){}
#[ cfg( all( feature = "enabled", feature = "type_variadic_from" ) )]
fn main()
{
  use variadic_from::exposed::*;

  // Define a struct `MyStruct` with fields `a` and `b`.
  // The struct derives common traits like `Debug`, `PartialEq`, `Default`
  // `VariadicFrom` defined manually.
  #[ derive( Debug, PartialEq, Default ) ]
  struct MyStruct
  {
    a : i32,
    b : i32,
  }

  // Implement the `From1` trait for `MyStruct`, which allows constructing a `MyStruct` instance
  // from a single `i32` value by assigning it to both `a` and `b` fields.
  impl From1< i32 > for MyStruct
  {
    fn from1( a : i32 ) -> Self { Self { a, b : a } }
  }

  // == begin of generated

  impl From2< i32, i32 > for MyStruct
  {
    fn from2( a : i32, b : i32 ) -> Self { Self{ a : a, b : b } }
  }

  impl From< ( i32, i32 ) > for MyStruct
  {
    #[ inline( always ) ]
    fn from( ( a, b ) : ( i32, i32 ) ) -> Self
    {
      Self::from2( a, b )
    }
  }

  // == end of generated

  let got : MyStruct = from!();
  let exp = MyStruct { a : 0, b : 0 };
  assert_eq!( got, exp );

  let got : MyStruct = from!( 13 );
  let exp = MyStruct { a : 13, b : 13 };
  assert_eq!( got, exp );

  let got : MyStruct = from!( 13, 14 );
  let exp = MyStruct { a : 13, b : 14 };
  assert_eq!( got, exp );

  dbg!( exp );
  //> MyStruct {
  //>   a : 13,
  //>   b : 14,
  //> }

}
```

</details>

Try out `cargo run --example variadic_from_trivial`.
<br/>
[See code](./examples/variadic_from_trivial.rs).

### To add to your project

```sh
cargo add variadic_from
```

### Try out from the repository

```sh
git clone https://github.com/Wandalen/wTools
cd wTools
cargo run --example variadic_from_trivial
```
