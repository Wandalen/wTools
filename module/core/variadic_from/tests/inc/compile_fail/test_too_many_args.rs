use variadic_from::from;

fn main()
{
  let _ = from!( 1, 2, 3, 4 );
}