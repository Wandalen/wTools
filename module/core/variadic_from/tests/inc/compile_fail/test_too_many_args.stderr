error: Too many arguments
 --> tests/inc/compile_fail/test_too_many_args.rs:5:11
  |
5 |   let _ = from!( 1, 2, 3, 4 );
  |           ^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `from` (in Nightly builds, run with -Z macro-backtrace for more info)
